### PR TITLE
Little improvements on header/source layout and embracing better practices

### PR DESCRIPTION
**Pull request by cthulhu-irl**
The changes are:
 - move `include/touca/cli/` to `cli/touca/cli/` since it's internal header directory and not part of library.
 - move `src/client/touca.cpp` to `src/touca.cpp` as it is the implementation of `include/touca/touca.hpp`. (authentical directory layout between `src/` and `include/touca/`)
 - rename `sink` type in `include/touca/core/variant.hpp` to `Discarder`, and set its assignment operator to delete.
 - rename `generic_overload` in `include/touca/core/variant.hpp` to `GenericOverload`.
 - make internal functions in `src/runner/options.cpp` static.
 - make internal functions in `src/runner/runner.cpp` static.
 - add virtual deconstructor to `touca::logger` interface class.
 - use if-else clauses rather than std::unordered_map and for-loop in `parse_env_variables` function in `src/client/options.cpp`.
 - use switch-case rather than std::unordered_map in Sink::LogLevel enum `stringify` function in `src/runner/runner.cpp`.

### CODE REVIEW

File: sdk/cpp/src/client/options.cpp [hunk 1]
```diff
@@ -32,17 +32,14 @@ func_t parse_member(bool& member) {
...
-    }
-  }
+  if (const auto env_value = std::getenv("TOUCA_API_KEY"))
+    options.api_key = env_value;
+
+  if (const auto env_value = std::getenv("TOUCA_API_URL"))
+    options.api_url = env_value;
+
+  if (const auto env_value = std::getenv("TOUCA_TEST_VERSION"))
+    options.revision = env_value;
```
Comment (ghorbanzade) [inline sdk/cpp/src/client/options.cpp:42]: We use braces for single-statement conditions.
```suggestion
  if (const auto value = std::getenv("TOUCA_API_KEY")) {
    options.api_key = value;
  }
  if (const auto value = std::getenv("TOUCA_API_URL")) {
    options.api_url = value;
  }
  if (const auto value = std::getenv("TOUCA_TEST_VERSION")) {
    options.revision = value;
  }
```

File: sdk/cpp/src/runner/runner.cpp [hunk 2]
```diff
@@ -163,14 +163,18 @@ void Logger::publish(const Sink::Level level, const std::string& msg) const {
...
+    case Sink::Level::Debug:
+      return "debug";
+    case Sink::Level::Info:
+      return "info";
+    case Sink::Level::Warn:
+      return "warning";
+    case Sink::Level::Error:
+      return "error";
+  }
+  return "";
```
Comment (ghorbanzade) [inline sdk/cpp/src/runner/runner.cpp:177]: Shouldn't we throw `std::invalid_argument` instead?

**Review by ghorbanzade — CHANGES_REQUESTED** [review]
Thank you for the PR!

File: sdk/cpp/src/runner/runner.cpp (hunk 2 above)
Comment (cthulhu-irl) [reply sdk/cpp/src/runner/runner.cpp:177]: it's unreachable, but yeah, I agree that it's better to throw that exception in case of UB.

**Review by ghorbanzade — APPROVED** [review]

**Pull request merged**